src/mlpack/methods/preprocess/preprocess_string_main.cpp

```diff
@@ -0,0 +1,301 @@
+                 const std::set<size_t>& dimension)
+{
+  ofstream fout;
+  fout.open(outputFilename, ios::trunc);
+  if (!fout.is_open())
+  {
+    Log::Fatal << "Unable to open a file for writing output.\n";
+  }
+  std::set<size_t>::iterator dimensionIterator;
+  for (size_t i = 0 ; i < dataset.size(); i++)
+  {
+    size_t col = 0;
+    dimensionIterator = dimension.begin();
+    for (size_t j = 0 ; j < dataset[i].size(); j++)
+    {
+      if (col < dimension.size() && *dimensionIterator == j)
+      {
+        if (j + 1 < dataset[i].size())
+          fout<<nonNumericInput[col][i]<<column_delimiter;
+        else
+          fout<<nonNumericInput[col][i];
+        dimensionIterator++;
+        col++;
+      }
+      else
+      {
+        if (j + 1 < dataset[i].size())
+          fout<<dataset[i][j]<<column_delimiter;
+        else
+          fout<<dataset[i][j];
+      }
+    }
+    fout<<"\n";
+  } 
+}
+
+static void mlpackMain()
+{
+  // Parse command line options.
+  // Extracting the filename
+  const std::string filename = CLI::GetParam<std::string>("actual_dataset");
+  // This is very dangerous, Let's add a check tommorrow.
+  std::string column_delimiter;
+  if (CLI::HasParam("column_delimiter"))
+  {
+    column_delimiter = CLI::GetParam<std::string>("column_delimiter");
+    // Allow only 3 delimiters.
+    RequireParamValue<std::string>("column_delimiter", [](std::string del)
+        { return del == "\t" || del == "," || del == " "; }, true,
+        "Delimiter should be either \\t (tab) or , (comma) or ' ' (space) ");
+  }
+  else
+  {
+      if (data::Extension(filename) == "csv")
+      {
+        column_delimiter = ",";
+        Log::Warn << "Found csv Extension, taking , as column_delimiter. \n";
+      }
+      else if (data::Extension(filename) == "tsv" ||
+          data::Extension(filename) == "txt")
+      {
+        column_delimiter = "\t";
+        Log::Warn << "Found tsv or txt Extension, taking \\t as"
+        "column_delimiter. \n";
+      }
+      else
+      {
+        column_delimiter = "\t";
+        Log::Warn << "column_delimiter not specified, taking default value \n";
+      }
+  }
+  // Handling Dimension vector
+  std::vector<std::string> temp_dimension =
+      CLI::GetParam<std::vector<std::string> >("dimension");
+  std::set<size_t> dimension;
+  getColumn(temp_dimension, dimension);
+  std::vector<std::vector<std::string>> dataset;
+  createDataset(dataset, filename, column_delimiter);
+  // Preparing the input dataset on which string manipulation has to be done.
+  std::vector<std::vector<std::string>> nonNumericInput(dimension.size());
+  std::set<size_t>::iterator dimensionIterator = dimension.begin();
+  for (size_t i = 0; i < dataset.size(); i++)
+  {
+    size_t col = 0;
+    dimensionIterator = dimension.begin();
+    for (size_t j = 0; j < dataset[i].size(); j++)
+    {
+      if (col < dimension.size())
+      {
+        if (*dimensionIterator == j)
```

Comment:
Again, it's too complicated. I suggest to use `std::unordered_set` and `find()`.

Reply:
Suppose we have 50 odd columns,
And the set has column 1,6,7 and then

I can break of after 7 th column
So potentially it's row*7

Where in case of unordered_map , I will have to iterate over the column full

I guess 
I can just iterate over the coulmn and insert then value rather than checking 

So I can do iterate over the set 
And just insert 
`nonNumericImput[col].push_back(dataset [I][set_elem]

Reply:
I will use unordered_map :)

Reply:
You don't need to iterate over the set. You have already iterated over the whole column set:
```c++
for (size_t j = 0; j < dataset[i].size(); j++)
```

Reply:
I see. Probably its a good idea to store the column index and the column data inside the same container:
```c++
vector<pair<size_t, vector<string>>> nonNumericInput;
```
or
```c++
unordered_map<size_t, vector<string>> nonNumericInput;
```
In that case you needn't iterate over the whole column set.